doc/developer/design/20210622_progress_observation.md

```diff
@@ -0,0 +1,104 @@
+# Observe source progress
+
+## Summary
+
+Currently, Materialize does not provide any feedback about the source
+progression. This makes it difficult to determine if a source is up-to-date and
+queries return fresh data. The goal of this effort is to add a mechanism to
+allow clients to query the progress of a source in a provider-independent way.
+
+## Goals
+
+The goals of this design are to enable users to observe the progress of their
+sources. It should help answer the following questions:
+
+* Is all (past) data loaded?
+* What was the last time the source was up-to-date?
+* How much does a source lag behind?
```

Comment:
I think we can tighten the goals of this a bit: My read of issue #4987 (and personal desire for stuff like active-instance selection a la blue/green) is that we only care about how behind the source is in relation to the state at the time the source was created. That is, sources can be in one of these two states:

* known-behind (because it's not yet consumed to the offset/object/etc that we knew about at source-creation time)
* caught up (to the point of minimal usefulness), aka "live"

I think it would be useful to know if a "live" source is falling behind, but maybe we can exclude that for now: The date we really care about is whether the source is as the "minimally useful" state yet.

Reply:
All that is to say, I think this is missing language around created-and-catching-up sources (: